src/mesh/Utils.cpp

```diff
@@ -0,0 +1,78 @@
+#include <Eigen/Core>
+#include <mesh/Edge.hpp>
+#include <mesh/Mesh.hpp>
+#include <utils/MasterSlave.hpp>
+
+namespace precice {
+namespace mesh {
+
+/// Given the data and the mesh, this function returns the surface integral. Assumes no overlap exists for the mesh
+Eigen::VectorXd integrate(PtrMesh mesh, PtrData data)
+{
+  const int       valueDimensions = data->getDimensions();
+  const int       meshDimensions  = mesh->getDimensions();
+  const auto &    values          = data->values();
+  Eigen::VectorXd integral        = Eigen::VectorXd::Zero(valueDimensions);
+
+  if (meshDimensions == 2) {
+    for (const auto &edge : mesh->edges()) {
+      int vertex1 = edge.vertex(0).getID() * valueDimensions;
+      int vertex2 = edge.vertex(1).getID() * valueDimensions;
+      for (int dim = 0; dim < valueDimensions; ++dim) {
+        integral(dim) += 0.5 * edge.getLength() * (values(vertex1 + dim) + values(vertex2 + dim));
+      }
+    }
+  } else {
+    for (const auto &face : mesh->triangles()) {
+      int vertex1 = face.vertex(0).getID() * valueDimensions;
+      int vertex2 = face.vertex(1).getID() * valueDimensions;
+      int vertex3 = face.vertex(2).getID() * valueDimensions;
+
+      for (int dim = 0; dim < valueDimensions; ++dim) {
+        integral(dim) += (face.getArea() / 3.0) * (values(vertex1 + dim) + values(vertex2 + dim) + values(vertex3 + dim));
+      }
+    }
+  }
+  return integral;
+}
+
+/// Given the data and the mesh, this function returns the surface integral
+Eigen::VectorXd integrateOwner(PtrMesh mesh, PtrData data)
+{
+  const int       valueDimensions = data->getDimensions();
+  const int       meshDimensions  = mesh->getDimensions();
+  const auto &    values          = data->values();
+  Eigen::VectorXd integral        = Eigen::VectorXd::Zero(valueDimensions);
+
+  if (meshDimensions == 2) {
+    for (const auto &edge : mesh->edges()) {
+      int vertex1 = edge.vertex(0).getID() * valueDimensions;
+      int vertex2 = edge.vertex(1).getID() * valueDimensions;
+
+      if (edge.vertex(0).isOwner() and edge.vertex(1).isOwner()) {
+        for (int dim = 0; dim < valueDimensions; ++dim) {
+          integral(dim) += 0.5 * edge.getLength() * (values(vertex1 + dim) + values(vertex2 + dim));
+        }
+      }
+    }
+  } else {
+    for (const auto &face : mesh->triangles()) {
+      int vertex1 = face.vertex(0).getID() * valueDimensions;
+      int vertex2 = face.vertex(1).getID() * valueDimensions;
+      int vertex3 = face.vertex(2).getID() * valueDimensions;
+
+      // If the rank owns at least two of the vertices of the triangle, calculate the integral
+      if ((face.vertex(0).isOwner() + face.vertex(1).isOwner() + face.vertex(2).isOwner()) > 1) {
```

Comment:
All three vertices could have different owners. Similar as above we could take 1/3 per owned vertex.

Reply:
I give a little bit of thought to this idea but I think the problem would still persist. Lets' have an example (it is taken from one of the tests), the values in parenthesis are the ownership information.

Rank 0 defines a triangle with vertices: 0 1 2 (1, 1, 0)
Rank 1 defines a triangle with vertices: 0 1 2 (0, 0, 1)
Rank 1 defines a triangle with vertices: 3 0 2 (1, 0, 1)

If we use a ratio based on the ownership of the vertices, it will calculate the integral correctly for the (0 1 2) triangle, but not for the (3 0 2) triangle, since we multiply it with 2/3 which is wrong. I think the fundamental reason why this is complicated is that we not know how many ranks define a triangle. Maybe the ownership information for edge and triangle would also be useful, I have no idea whether it can be implemented or not though. What do you think about this?

Reply:
Yes, you are completely right. Would have been too easy to be true. Yes, I agree that for a proper solution we would need sth like ownership for edges and triangles. That's not impossible to do. I guess a gather-scatter solution in the following function should not be too complicated to do:
https://github.com/precice/precice/blob/00589fe951f9715174eafbd66657294e0d8eb8e4/src/partition/ReceivedPartition.cpp#L468